File: 2022/day-14/day14.py

import re
from typing import *
from math import copysign
from operator import itemgetter as item

Point = Tuple[int, int]
Rocks = Set[Point]
Board = Tuple[Rocks, Point, Point]
FinishingCondition = Callable[[Board, Point], bool]


def add(a: Point,  b: Point) -> Point:
    ar, ad = a
    br, bd = b
    return br + ar, bd + ad


def sub(a: Point, b: Point) -> Point:
    br, bd = b
    return add(a, (-br, -bd))


def clamp(a: int) -> int:
    return int(copysign(1, a))


def unit_vector(a: Point) -> Point:
    ar, ad = a
    return clamp(ar) if ar != 0 else 0, clamp(ad) if ad != 0 else 0


def make_line(start: Point, end: Point) -> Rocks:
    rocks = {start, end}
    p = start
    d = unit_vector(sub(end, start))
    while p != end:
        p = add(p, d)
        rocks.add(p)
    return rocks


def parse_input(inp: str, part_two=False) -> Board:
    point_regex = re.compile(r"(\d+),(\d+)")
    rocks: Rocks = set()

    for line in inp.splitlines():
        l = [tuple(map(int, m.groups())) for m in point_regex.finditer(line)]
        starts, ends = l[:-1], l[1:]
        for start, end in zip(starts, ends):
            rocks |= make_line(start, end)

    _, max_down = max(rocks, key=item(1))
    if part_two:
        max_down += 2
        # bottom line does not need to be 2 * max_down
        rocks |= make_line((500-max_down-5, max_down), (500 + max_down + 5, max_down))

    min_right, _ = min(rocks, key=item(0))
    max_right, _ = max(rocks, key=item(0))

    return rocks, (min_right, 0), (max_right, max_down)


def fell_into_abyss(board: Board, _: Point, new_pos: Optional[Point]) -> bool:
    if new_pos is None:
        return False
    _, _, (_, ld) = board
    _, pd = new_pos
    return pd > ld


def source_reached(_: Board, grain: Point, new_pos: Optional[Point]) -> bool:
    return grain == (500, 0) and new_pos is None


def simulate(board: Board, source: Point, finished: FinishingCondition) -> Rocks:
    rocks, _, (_, ld) = board
    sand = set()

    def can_move(gr: Point) -> Optional[Point]:
        directions = [(0, 1), (-1, 1), (1, 1)]
        for direction in directions:
            new_pos = add(gr, direction)
            if new_pos not in rocks and new_pos not in sand:
                return new_pos
        return None

    can_add_sand = True
    while can_add_sand:
        grain = source
        p = can_move(grain)
        can_add_sand = not finished(board, grain, p)
        while p is not None:
            grain = p
            p = can_move(grain)
            can_add_sand = not finished(board, grain, p)
            if not can_add_sand:
                break
        if can_add_sand:
            sand.add(grain)
        else:
            break
    return sand


with open('input.txt') as h:
     content = h.read()

source = (500, 0)

game_board = parse_input(content)
print('Part 1: ', len(simulate(game_board, source, fell_into_abyss)))
game_board = parse_input(content, part_two=True)
print('Part 2: ', len(simulate(game_board, source, source_reached))+1)

File: 2022/day-14/day14.rs
use std::collections::HashSet;

type Point = (i32, i32);
type Rocks = HashSet<Point>;
type Board = (Rocks, Point, Point);

fn point_op<T: Clone>(a: &(T, T), b: &(T, T), op: impl Fn(T, T) -> T) -> (T, T) {
    let (ar, ad) = a.clone();
    let (br, bd) = b.clone();
    (op(ar, br), op(ad, bd))
}

fn add(a: &Point, b: &Point) -> Point {
    point_op(a, b, std::ops::Add::add)
}

fn sub(a: &Point, b: &Point) -> Point {
    point_op(a, b, std::ops::Sub::sub)
}

fn unit_vector(a: Point) -> Point {
    (
        if a.0 != 0 { a.0.signum() } else { 0 },
        if a.1 != 0 { a.1.signum() } else { 0 },
    )
}

fn make_line(start: &Point, end: &Point) -> Rocks {
    let mut rocks: Rocks = HashSet::from_iter(vec![start.clone(), end.clone()]);
    let diff = unit_vector(sub(end, start));
    let mut p = *start;
    while &p != end {
        p = add(&p, &diff);
        rocks.insert(p.clone());
    }
    rocks
}

fn parse_input(contents: &str, part_two: bool) -> Board {
    let lines: Vec<Vec<Point>> = contents
        .lines()
        .map(|l| {
            l.split(" -> ")
                .map(|coords| {
                    let mut sep = coords.split(',');
                    (
                        sep.next().unwrap().parse().unwrap(),
                        sep.next().unwrap().parse().unwrap(),
                    )
                })
                .collect()
        })
        .collect();
    let mut rocks = Rocks::new();
    lines.iter().for_each(|line| {
        line.iter()
            .take(line.len() - 1)
            .zip(line.iter().skip(1))
            .for_each(|(start, end)| {
                make_line(start, end).iter().for_each(|&p| {
                    rocks.insert(p);
                })
            })
    });
    let mut max_down = rocks.iter().map(|(_, d)| *d).max().unwrap();

    if part_two {
        max_down += 2;
        // bottom line does not need to be 2 * max_down
        make_line(
            &(500 - max_down - 1, max_down),
            &(500 + max_down + 1, max_down),
        )
        .iter()
        .for_each(|&p| {
            rocks.insert(p);
        });
    }
    let max_right = rocks.iter().map(|(r, _)| *r).max().unwrap();
    let min_point = (rocks.iter().map(|(r, _)| *r).min().unwrap(), 0);

    (rocks, min_point, (max_right, max_down))
}

fn fell_into_abyss(board: &Board, _: &Point, new_pos: Option<Point>) -> bool {
    match new_pos {
        Some((_, pd)) => {
            let (_, _, (_, ld)) = board;
            pd > *ld
        }
        None => false,
    }
}

fn source_reached(_: &Board, grain: &Point, new_pos: Option<Point>) -> bool {
    grain == &(500, 0) && new_pos.is_none()
}

fn can_move(gr: &Point, rocks: &Rocks, sand: &Rocks) -> Option<Point> {
    for direction in &[(0, 1), (-1, 1), (1, 1)] {
        let new_pos = add(&gr, direction);
        if !rocks.contains(&new_pos) && !sand.contains(&new_pos) {
            return Some(new_pos);
        }
    }
    None
}

fn simulate(
    board: &Board,
    source: &Point,
    finished: impl Fn(&Board, &Point, Option<Point>) -> bool,
) -> Rocks {
    let (rocks, _, _) = board;
    let mut sand = Rocks::new();

    let mut can_add_sand = true;
    while can_add_sand {
        let mut grain = source.clone();
        let mut p = can_move(&grain, rocks, &sand);
        can_add_sand = !finished(board, &grain, p);
        while p.is_some() {
            grain = p.unwrap();
            p = can_move(&grain, rocks, &sand);
            can_add_sand = !finished(board, &grain, p);
            if !can_add_sand {
                break;
            }
        }
        if can_add_sand {
            sand.insert(grain.clone());
        } else {
            break;
        }
    }
    return sand;
}

fn draw_board(board: &Board) {
    let (rocks, minp, maxp) = board;
    for y in minp.1..=maxp.1 {
        for x in minp.0..=maxp.0 {
            let mut ch = if rocks.contains(&(x, y)) { '#' } else { '.' };
            print!("{}", ch);
        }
        println!("");
    }
}

fn main() {
    let contents = std::fs::read_to_string("input.txt").expect("Failed to read input!");
    let source = (500, 0);
    let board = parse_input(&contents, false);
    println!(
        "Part 1: {}",
        simulate(&board, &source, fell_into_abyss).len()
    );
    let board = parse_input(&contents, true);
    println!(
        "Part 2: {}",
        simulate(&board, &source, source_reached).len()
    );
}
